lxd/storage.go

```diff
@@ -387,6 +387,89 @@ func storagePoolInit(d *Daemon, poolName string) (storage, error) {
 	return storageInit(d, poolName, "", -1)
 }
 
+func storagePoolVolumeAttachInit(d *Daemon, poolName string, volumeName string, volumeType int, idmapset *shared.IdmapSet) (storage, error) {
+	st, err := storageInit(d, poolName, volumeName, volumeType)
+	if err != nil {
+		return nil, err
+	}
+
+	var lastIdmap *shared.IdmapSet
+	poolVolumePut := st.GetStoragePoolVolumeWritable()
+	if poolVolumePut.Config["volatile.idmap"] != "" {
+		// Convert the volume type name to our internal integer representation.
+		volumeTypeName, err := storagePoolVolumeTypeToName(volumeType)
+		if err != nil {
+			return nil, err
+		}
+
+		volumeUsedBy, err := storagePoolVolumeUsedByGet(d, volumeName, volumeTypeName)
+		if err != nil {
+			return nil, err
+		}
+
+		lastIdmap = new(shared.IdmapSet)
+		err = json.Unmarshal([]byte(poolVolumePut.Config["volatile.idmap"]), &lastIdmap)
+		if err != nil {
+			return nil, err
+		}
+
+		if !reflect.DeepEqual(idmapset, lastIdmap) {
+			if len(volumeUsedBy) > 0 {
+				return nil, fmt.Errorf("idmaps are not identical")
+			}
+		} else {
+			return st, nil
+		}
+	}
+
+	idmapBytes, err := json.Marshal(idmapset)
+	if err != nil {
+		return nil, err
+	}
+	jsonIdmap := string(idmapBytes)
+	poolVolumePut.Config["volatile.idmap"] = jsonIdmap
+
+	st.SetStoragePoolVolumeWritable(&poolVolumePut)
+
+	err = storagePoolVolumeUpdate(d, poolName, volumeName, volumeType, poolVolumePut.Description, poolVolumePut.Config)
+	if err != nil {
+		return nil, err
+	}
+
+	ourMount, err := st.StoragePoolVolumeMount()
+	if err != nil {
+		return nil, err
+	}
+	if ourMount {
+		defer func () {
+			_, err := st.StoragePoolVolumeUmount()
+			if err != nil {
+				logger.Warnf("failed to unmount storage volume")
+			}
+		}()
+	}
+
+	remapPath := getStoragePoolVolumeMountPoint(poolName, volumeName)
+
+	if lastIdmap != nil {
+		err := lastIdmap.UnshiftRootfs(remapPath)
+		if err != nil {
+			return nil, err
+		}
+		logger.Debugf("shifted storage volume")
+	}
+
+	if idmapset != nil {
+		err := idmapset.ShiftRootfs(remapPath)
+		if err != nil {
+			return nil, err
+		}
+		logger.Debugf("shifted storage volume")
+	}
+
+	return st, nil
```

Comment:
Np, I just jotted this down in an hour so we could align and land this tomorrow. :)